proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+
+### Non-Goals
+
+- Define how NCP controller works.
+
+- Discuss the details of how the third-party component provider work.
+
+## Proposal
+
+### Portability and Customizability
+
+Generally, creating the three major components requires similar high-level information, like the components' version, the number of replicas, and the amount of computing resources. Meanwhile, end-users should be able to customize NCP components, i.e., specifying the component image, version, and command-line options. Therefore, we define a new struct `NestedComponentSpec` that contains common information required by different providers as well as customized information specified by the end-users. The `NestedComponentSpec` will look like the following
+
+```go
+type NestedComponentSpec struct {
+    // ComponentSpec defines the common information for creating the component
+    // +optional
+    CommonSpec addonv1alpha1.CommonSpec `json:"commonSpec",omitempty`
+    
+    // Resources defines the amount of computing resources that will be used by this component
+    // +optional
+    Resources corev1.ResourceRequirements `json:"resources",omitempty`
+    
+    // Replicas defines the number of replicas in the component's workload 
+    // +optional
+    Replicas int32 `json:"replicas",omitempty`
+
+    // PatchSpecs includes the user specifed settings
+    // +optional
+    PatchSpec addonv1alpha1.PatchSpec `json:"patchSpec,omitempty"`
+}
+```
+
+The `CommonSpecs` and the `PatchSpec` are defined in [kubebuilder-declarative-pattern](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/1cbf859290cab81ae8e73fc5caebe792280175d1/pkg/patterns/addon/pkg/apis/v1alpha1/common_types.go):
+
+
+```go
+// CommonSpec defines the set of configuration attributes that must be exposed on all addons.
+type CommonSpec struct {
+	// Version specifies the exact addon version to be deployed, eg 1.2.3
+	// It should not be specified if Channel is specified
+	Version string `json:"version,omitempty"`
+	// Channel specifies a channel that can be used to resolve a specific addon, eg: stable
+	// It will be ignored if Version is specified
+	Channel string `json:"channel,omitempty"`
+}
+
+// +k8s:deepcopy-gen=true
+type PatchSpec struct {
+	Patches []*runtime.RawExtension `json:"patches,omitempty"`
+}
+```
+
+If the designated component's providers do not allow patching, the `PatchSpec` will be ignored. All Nested CRs need to contain a `NestedComponentSpec`. We will not discuss how NCPT and NCP controller generates the `NestedComponentSpec` in this proposal. However, a possible way is to add three fields to the NCPT CRD, and has each of them holds the `NestedComponentSpec` for one of the components. The NCPT will look like the following
+
+```go
+// NestedControlPlaneTemplateSpec defines the desired state of NestedControlPlaneTemplate
+type NestedControlPlaneTemplateSpec struct {
```

Comment:
`Template` may not be a good name here, not sure.

But I personally like the idea in VC where we have two different concepts: `ClusterVersions` CR and `VirtualCluster` CR.
This easily addresses 3 major use cases in daily operations, for example:
1. The super admin has the responsibility/permission to maintain the "available versions of the Control Plane" to be used by tenant admins -- btw, we do allow the tenant admins to spin up new tenant masters, don't we?
2. The tenant admin spins up tenant clusters by simply referring to any of the "available versions of the Control Plane" -- just like claiming PVC against a predefined PV, I guess.
3. The super admin has the responsibility/permission to deprecate older/vulnerable versions to avoid too much flexibility granted to tenant admins if they can literarily define any K8s version to build the tenant masters -- this sounds like giving a way to tenant admins to hack into the super cluster -- at least it's not a good idea to maintain the manageability. What do you guys think?

Reply:
@brightzheng100 I agree that `template` can be helpful to manage different settings. But in the first version, we may just implement something more generic/flexible, i.e., creating each CR independently, and we can build the `template` mechanism on top of it later.

Reply:
Is the template concept going to be implemented by CAPN itself? Or is this something that we're expecting upstream Cluster API to implement?

Reply:
Agree @brightzheng100 and @charleszheng44. These are all functions that can layer later to make the DX improved but would be hard to treat as a two-way door if we forced that style first.

@vincepri I believe that will still end up in CAPN as a way to augment and make CAPN easier to consume for some users  like Bright has brought up, which is where VC makes the DX really simple but forces all implementations to implement copies and makes upgrades somewhat complex because there isn't a single owner of each resource to maintain updating of that component.